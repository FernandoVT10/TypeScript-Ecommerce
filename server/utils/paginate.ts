export interface IPagination {
    hasPrevPage: boolean,
    prevPage: number,
    hasNextPage: boolean,
    nextPage: number,
    pages: {
        active: boolean,
        pageNumber: number
    }[]
}

export default function paginate(totalPages: number, cursor: number): IPagination {
    let remainingLinks = 6;
    let minPage = 0, maxPage = 0;

    if(cursor + 3 > totalPages) {
        remainingLinks -= totalPages - cursor;
        maxPage = totalPages;
    } else {
        remainingLinks -= 3;
        
        if(cursor < 4) {
            if(totalPages > 7) {
                maxPage = 7;
            } else {
                maxPage = totalPages;
            }
        } else {
            maxPage = cursor + 3;
        }
    }

    if(cursor - remainingLinks > 0) {
        minPage = cursor - remainingLinks;
    } else {
        minPage = 1;
    }

    const result: IPagination = {
        hasPrevPage: false,
        prevPage: null,
        hasNextPage: false,
        nextPage: null,
        pages: []
    };

    if(cursor > 1) {
        result.hasPrevPage = true;
        result.prevPage = cursor - 1;
    }

    if(cursor < totalPages) {
        result.hasNextPage = true;
        result.nextPage = cursor + 1;
    }

    for(let i = minPage; i <= maxPage; i++) {
        if(i === cursor) {
            result.pages.push({
                active: true,
                pageNumber: i
            });
        } else {
            result.pages.push({
                active: false,
                pageNumber: i
            });
        }
    }

    return result;
}