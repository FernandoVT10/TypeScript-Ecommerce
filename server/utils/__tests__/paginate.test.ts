import paginate from "../paginate";

describe("Utils Paginate", () => {
    it("should get 7 pages with the 'hasPrevPage' and 'hasNextPage' active", () => {
        const pagination = paginate(100, 72);

        expect(pagination).toMatchSnapshot();
    });

    it("should get the 'hasPrevPage' and 'hasNextPage' desactive", () => {
        const pagination = paginate(1, 1);

        expect(pagination.pages.length).toBe(1);
        expect(pagination.hasPrevPage).toBeFalsy();
        expect(pagination.hasNextPage).toBeFalsy();
    });

    it("should get the 'hasPrevPage' active", () => {
        const pagination = paginate(10, 10);

        expect(pagination.pages.length).toBe(7);
        expect(pagination.hasPrevPage).toBeTruthy();
        expect(pagination.hasNextPage).toBeFalsy();
    });

    it("should get the 'hasNextPage' active", () => {
        const pagination = paginate(10, 1);

        expect(pagination.pages.length).toBe(7);
        expect(pagination.hasPrevPage).toBeFalsy();
        expect(pagination.hasNextPage).toBeTruthy();
    });
});