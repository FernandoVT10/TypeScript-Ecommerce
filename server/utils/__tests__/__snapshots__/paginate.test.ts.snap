// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils Paginate should get 7 pages with the 'hasPrevPage' and 'hasNextPage' active 1`] = `
Object {
  "hasNextPage": true,
  "hasPrevPage": true,
  "nextPage": 73,
  "pages": Array [
    Object {
      "active": false,
      "pageNumber": 69,
    },
    Object {
      "active": false,
      "pageNumber": 70,
    },
    Object {
      "active": false,
      "pageNumber": 71,
    },
    Object {
      "active": true,
      "pageNumber": 72,
    },
    Object {
      "active": false,
      "pageNumber": 73,
    },
    Object {
      "active": false,
      "pageNumber": 74,
    },
    Object {
      "active": false,
      "pageNumber": 75,
    },
  ],
  "prevPage": 71,
}
`;
